src/btree/bt_curnext.c

```diff
@@ -750,13 +754,14 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 total_skipped += skipped;
                 break;
             case WT_PAGE_ROW_LEAF:
-                ret = __cursor_row_next(cbt, newpage, restart, &skipped, prefix);
+                ret = __cursor_row_next(cbt, newpage, restart, &prefix_key_out_of_bounds, &skipped, prefix);
                 total_skipped += skipped;
                 /*
-                 * We can directly return WT_NOTFOUND here as the caller expects the cursor to be
-                 * positioned when traversing keys for prefix search near.
+                 * If we are doing a search near with prefix key configured, we need to check if we
+                 * have exited the cursor row next function due to a prefix key mismatch. If so, we
+                 * can immediately return WT_NOTFOUND and we do not have to walk onto the next page.
                  */
-                if (ret == WT_NOTFOUND && F_ISSET(&cbt->iface, WT_CURSTD_PREFIX_SEARCH))
+                if (ret == WT_NOTFOUND && prefix_key_out_of_bounds)
```

Comment:
Do we still need to check for `ret == WT_NOTFOUND`? We could assert if we wanted to keep it but given where `prefix_key_out_of_bounds` is set to true, ret should be `WT_NOTFOUND`.

Reply:
I agree with an assert here.

Reply:
Ok, I will probably use a WT_ASSERT here instead.